This program uses Dijkstra's algorithm to find the shortest route
from Hyderabad to Bengaluru using a weighted graph of Indian cities.

In [1]:
# Dijkstra's algorithm for finding the shortest path between Indian cities

graph = {
    "Hyderabad": {
        "Kurnool": 210,
        "Pune": 560,
        "Nagpur": 500
    },

    "Kurnool": {
        "Hyderabad": 210,
        "Bengaluru": 360,
        "Chennai": 380
    },

    "Pune": {
        "Hyderabad": 560,
        "Bengaluru": 840,
        "Nagpur": 710
    },

    "Nagpur": {
        "Hyderabad": 500,
        "Pune": 710
    },

    "Chennai": {
        "Kurnool": 380,
        "Bengaluru": 350
    },

    "Bengaluru": {
        "Kurnool": 360,
        "Pune": 840,
        "Chennai": 350
    }
}

start = "Hyderabad"
goal = "Bengaluru"

# Set initial distances
distances = {}

for city in graph:
    distances[city] = float("inf")

distances[start] = 0

# To store the previous city
previous = {}

visited = set()

# Dijkstra's algorithm
while len(visited) < len(graph):

    current = None

    for city in graph:
        if city not in visited:
            if current is None or distances[city] < distances[current]:
                current = city

    if current is None:
        break

    visited.add(current)

    # Stop when the goal is reached
    if current == goal:
        break

    # Update distances of connected cities
    for neighbour, distance in graph[current].items():

        new_distance = distances[current] + distance

        if new_distance < distances[neighbour]:
            distances[neighbour] = new_distance
            previous[neighbour] = current

# Find the path
path = []
current = goal

while current != start:
    path.append(current)
    current = previous[current]

path.append(start)
path.reverse()

print("Starting city:", start)
print("Goal city:", goal)
print("Shortest path:", " -> ".join(path))
print("Total distance:", distances[goal], "km")

Starting city: Hyderabad
Goal city: Bengaluru
Shortest path: Hyderabad -> Kurnool -> Bengaluru
Total distance: 570 km


## Results

The program begins from Hyderabad and finds out the shortest path to Bengaluru.

There can be various paths to reach there, however, Dijkstra’s algorithm computes
both distances and chooses the one with the shortest distance.

The shortest path found out is:

Hyderabad → Kurnool → Bengaluru

The total distance is 570 km.